crypto/dilithium/p_dilithium_test.cc

```diff
@@ -21,368 +22,142 @@
 
-TEST(Dilithium3Test, KeyGeneration) {
-  // Basic key generation tests for Dilithium3
-  // Generate a Dilithium3 key
-  EVP_PKEY_CTX *dilithium_pkey_ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_DILITHIUM3, nullptr);
-  ASSERT_NE(dilithium_pkey_ctx, nullptr);
+#define CMP_VEC_AND_PTR(vec, ptr, len)         \
+{                                    \
+std::vector<uint8_t> tmp(len);     \
+tmp.assign(ptr, ptr+len);          \
+EXPECT_EQ(Bytes(vec), Bytes(tmp)); \
+}
+
+#define CMP_VEC_AND_PKEY_PUBLIC(vec, pkey, len) \
+CMP_VEC_AND_PTR(vec, pkey->pkey.pqdsa_key->public_key, len)
+
+#define CMP_VEC_AND_PKEY_SECRET(vec, pkey, len) \
+CMP_VEC_AND_PTR(vec, pkey->pkey.pqdsa_key->secret_key, len)
+
+static const struct ML_DSA parameterSet[] = {
+  {"MLDSA65", NID_MLDSA65, 1952, 4032, 3309,  "dilithium/kat/mldsa65.txt",mldsa65kPublicKey, mldsa65kPublicKeySPKI, 1973},
+};
+
+class MLDSAParameterTest : public testing::TestWithParam<ML_DSA> {};
+
+INSTANTIATE_TEST_SUITE_P(All, MLDSAParameterTest, testing::ValuesIn(parameterSet),
+                         [](const testing::TestParamInfo<ML_DSA> &params)
+                             -> std::string { return params.param.name; });
+
+
+TEST_P(MLDSAParameterTest, KAT) {
+  std::string kat_filepath = "crypto/";
+  kat_filepath += GetParam().kat_filename;
+
+  FileTestGTest(kat_filepath.c_str(), [&](FileTest *t) {
+    std::string count, mlen, smlen;
+    std::vector<uint8_t> seed, msg, pk, sk, sm;
+
+    ASSERT_TRUE(t->GetAttribute(&count, "count"));
+    ASSERT_TRUE(t->GetBytes(&seed, "seed"));
+    ASSERT_TRUE(t->GetAttribute(&mlen, "mlen"));
+    ASSERT_TRUE(t->GetBytes(&msg, "msg"));
+    ASSERT_TRUE(t->GetBytes(&pk, "pk"));
+    ASSERT_TRUE(t->GetBytes(&sk, "sk"));
+    ASSERT_TRUE(t->GetAttribute(&smlen, "smlen"));
+    ASSERT_TRUE(t->GetBytes(&sm, "sm"));
+
+    size_t pk_len = GetParam().public_key_len;
+    size_t sk_len = GetParam().secret_key_len;
+    size_t sig_len = GetParam().signature_len;
+
+    // The KAT files generated by the dilithium team use the optional APIs that
+    // create a signature for a message m and append the message to the end of
+    // the signature. We only want to bring the APIs that create and verify just
+    // the signature, therefore each signature is a constant
+    // DILITHIUM3_SIGNATURE_BYTES and we truncate the signed message down to a
+    // constant DILITHIUM3_SIGNATURE_BYTES.
+
+    std::vector<uint8_t> signature(sig_len);
+    sm.resize(sig_len);
 
-  EVP_PKEY *dilithium_pkey = EVP_PKEY_new();
-  ASSERT_NE(dilithium_pkey, nullptr);
+    // Convert string read from KAT to int
+    size_t mlen_int = std::stoi(mlen);
 
-  EXPECT_TRUE(EVP_PKEY_keygen_init(dilithium_pkey_ctx));
-  EXPECT_TRUE(EVP_PKEY_keygen(dilithium_pkey_ctx, &dilithium_pkey));
-  ASSERT_NE(dilithium_pkey->pkey.ptr, nullptr);
+    // Here we fix the DRBG (AES-CTR) so that we are able to seed it with the
+    // seed from the KAT (testing only)
+    pq_custom_randombytes_use_deterministic_for_testing();
```

Comment:
are we going to have derand api for PQDSA algorithms so that we can get rid of this?

Reply:
Yes! The de-rand API will come in a subsequent PR, we are so close now!